File: backend/app/routers/members.py
import uuid

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.dependencies import (
    require_admin_for_remove,
    require_admin_for_roles,
    require_member,
)
from app.db.session import get_async_session
from app.schemas.membership import MemberRead, MemberRoleUpdate
from app.services import members as member_service

router = APIRouter(prefix="/api/v1/tenants/{tenant_id}/members", tags=["members"])


@router.get("/", response_model=list[MemberRead])
async def list_members(
    ctx=Depends(require_member),
    session: AsyncSession = Depends(get_async_session),
):
    return await member_service.list_members(session, ctx.tenant.id)


@router.patch("/{user_id}/role", response_model=MemberRead)
async def assign_role(
    user_id: uuid.UUID,
    payload: MemberRoleUpdate,
    ctx=Depends(require_admin_for_roles),
    session: AsyncSession = Depends(get_async_session),
):
    return await member_service.assign_role(session, ctx.tenant.id, user_id, payload)


@router.delete("/{user_id}", status_code=204)
async def remove_member(
    user_id: uuid.UUID,
    ctx=Depends(require_admin_for_remove),
    session: AsyncSession = Depends(get_async_session),
):
    await member_service.remove_member(session, ctx.tenant.id, user_id)

File: frontend/src/lib/members.ts
// Shapes mirror the backend members API (app/schemas/membership.py).

import { authedGet, authedRequest } from "@/lib/api";
import type { Role } from "@/lib/permissions";

export type Member = {
  user_id: string;
  email: string;
  name: string | null;
  role: Role;
  created_at: string;
  github_login: string | null;
  github_link_status: string | null;
};

export function listMembers(tenantId: string): Promise<Member[]> {
  return authedGet<Member[]>(`/api/v1/tenants/${tenantId}/members/`);
}

export function assignMemberRole(
  tenantId: string,
  userId: string,
  role: Role,
): Promise<Member> {
  return authedRequest<Member>(
    "PATCH",
    `/api/v1/tenants/${tenantId}/members/${userId}/role`,
    { role },
  );
}

export function removeMember(tenantId: string, userId: string): Promise<null> {
  return authedRequest<null>("DELETE", `/api/v1/tenants/${tenantId}/members/${userId}`);
}
